File: src/lib/bungie/syncFreshness.ts
import { API_ERROR_CODES, ApiError } from "@/lib/api/errors";
import type { AppDatabase } from "@/lib/db/client";
import { getInventoryStatus } from "@/lib/db/repositories/inventoryRepository";
import type { DbUser } from "@/lib/db/types";
import type { BungieProfileClient } from "@/lib/bungie/types";
import type { EntityCache, ManifestService } from "@/lib/manifest/types/services";
import {
  syncUserInventory,
  SyncInProgressError,
  type SyncInventoryResult,
} from "@/lib/bungie/syncInventory";

/** Reuse inventory sync when last full sync is within this window (DBR-EQP-007). */
export const EQUIP_SYNC_FRESH_MS = 60_000;

export type SyncIfStaleResult = {
  synced: boolean;
  lastFullSyncAt: string | null;
  result?: SyncInventoryResult;
};

export function isInventoryFresh(
  lastFullSyncAt: string | null | undefined,
  nowMs = Date.now(),
  freshMs = EQUIP_SYNC_FRESH_MS,
): boolean {
  if (!lastFullSyncAt) return false;
  const syncedAt = Date.parse(lastFullSyncAt);
  if (Number.isNaN(syncedAt)) return false;
  const age = nowMs - syncedAt;
  return age >= 0 && age < freshMs;
}

export async function syncIfStale(
  db: AppDatabase,
  user: DbUser,
  accessToken: string,
  profileClient: BungieProfileClient,
  entityCache: EntityCache,
  manifest: ManifestService,
  manifestVersion: string,
  nowMs = Date.now(),
): Promise<SyncIfStaleResult> {
  const status = getInventoryStatus(db, user.id);
  if (isInventoryFresh(status?.lastFullSyncAt, nowMs)) {
    return { synced: false, lastFullSyncAt: status?.lastFullSyncAt ?? null };
  }

  try {
    const result = await syncUserInventory(
      db,
      user,
      accessToken,
      profileClient,
      entityCache,
      manifest,
      manifestVersion,
    );
    return { synced: true, lastFullSyncAt: result.lastFullSyncAt, result };
  } catch (error) {
    if (error instanceof SyncInProgressError) {
      throw new ApiError(
        API_ERROR_CODES.SYNC_BUSY,
        "Inventory sync already in progress; retry shortly",
        undefined,
        409,
      );
    }
    throw error;
  }
}

File: packages/bungie/lib/src/sync/sync_freshness.dart
import 'package:destiny2_db/destiny2_db.dart';

import '../profile/bungie_profile_client.dart';
import '../profile/equipment_bucket_lookup.dart';
import 'sync_inventory.dart';

/// Reuse inventory sync when last full sync is within this window (DBR-EQP-007).
///
/// Product: `EQUIP_SYNC_FRESH_MS = 60_000`.
const int kEquipSyncFreshMs = 60000;

/// Result of [syncIfStale].
class SyncIfStaleResult {
  const SyncIfStaleResult({
    required this.synced,
    required this.lastFullSyncAt,
    this.result,
  });

  final bool synced;
  final String? lastFullSyncAt;
  final SyncInventoryResult? result;
}

/// Whether [lastFullSyncAt] is within [freshMs] of [nowMs].
///
/// Returns false for null, unparseable timestamps, or future-dated stamps
/// that would yield negative age.
bool isInventoryFresh(
  String? lastFullSyncAt, {
  int? nowMs,
  int freshMs = kEquipSyncFreshMs,
}) {
  if (lastFullSyncAt == null || lastFullSyncAt.isEmpty) return false;
  final syncedAt = DateTime.tryParse(lastFullSyncAt);
  if (syncedAt == null) return false;
  final now = nowMs ?? DateTime.now().millisecondsSinceEpoch;
  final age = now - syncedAt.millisecondsSinceEpoch;
  return age >= 0 && age < freshMs;
}

/// Sync inventory only when last full sync is missing or older than 60s.
///
/// Production hosts MUST pass [equipmentBucketLookup] and/or
/// [equipmentBucketLookupBuilder] so vault/postmaster gear is stored (DART-050).
Future<SyncIfStaleResult> syncIfStale({
  required AppDatabase db,
  required int userId,
  required String accessToken,
  required BungieProfileClient profileClient,
  Map<int, int>? equipmentBucketLookup,
  EquipmentBucketLookupBuilder? equipmentBucketLookupBuilder,
  String? now,
  int? nowMs,
  InventoryBusyLock? lock,
}) async {
  final status = await getInventoryStatus(db, userId);
  final clock = nowMs ??
      (now != null
          ? DateTime.parse(now).millisecondsSinceEpoch
          : DateTime.now().millisecondsSinceEpoch);

  if (isInventoryFresh(status?.lastFullSyncAt, nowMs: clock)) {
    return SyncIfStaleResult(
      synced: false,
      lastFullSyncAt: status?.lastFullSyncAt,
    );
  }

  final result = await syncUserInventory(
    db: db,
    userId: userId,
    accessToken: accessToken,
    profileClient: profileClient,
    equipmentBucketLookup: equipmentBucketLookup,
    equipmentBucketLookupBuilder: equipmentBucketLookupBuilder,
    now: now ??
        DateTime.fromMillisecondsSinceEpoch(clock, isUtc: true)
            .toIso8601String(),
    lock: lock,
  );

  return SyncIfStaleResult(
    synced: true,
    lastFullSyncAt: result.lastFullSyncAt,
    result: result,
  );
}
